### PR TITLE
perf(source): use GAT and TAIT to impl zero cost async `SourceParser` trait

### PR DESCRIPTION
**Pull request by waruto210**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- use GAT and TAIT to impl zero cost async `SourceParser` trait

## Checklist

~- [] I have written necessary rustdoc comments~
~- [] I have added necessary unit tests and integration tests~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Comment by BugenZhao** [comment]
How about extracting the implementation to `xxx_inner` so that the diff can be cleaner? 🤔

**Comment by waruto210** [comment]
> How about extracting the implementation to `xxx_inner` so that the diff can be cleaner? 🤔

Good idea.

**Comment by waruto210** [comment]
> How about extracting the implementation to `xxx_inner` so that the diff can be cleaner? 🤔

Done, PTAL.

**Review by TennyZhuang — APPROVED** [review]

**Review by BugenZhao — APPROVED** [review]

File: src/source/src/parser/maxwell/json_parser.rs [hunk 1]
```diff
@@ -37,9 +38,8 @@ pub struct MaxwellEvent {
 #[derive(Debug)]
 pub struct MaxwellParser;
 
-#[async_trait::async_trait]
-impl SourceParser for MaxwellParser {
-    async fn parse(
+impl MaxwellParser {
+    async fn parse_inner(
```
Comment (BugenZhao) [inline src/source/src/parser/maxwell/json_parser.rs:42]:
```suggestion
    fn parse_inner(
```

**Pull request merged**